Battle Royale

congratulatio opened the pull request:
This implements the BTR game type, from reverse engineering it using the [proxy logs](https://dev.azure.com/airmash/proxylogs/_git/proxylogs?version=GBroom-region&path=/btr1).

I'm hosting a test server for it, which can be accessed via:
* https://test.airmash.online/#eu-btr1, or 
* https://starmash.test.airmash.online/#eu-btr1

It's mostly complete I reckon, just needs some review and testing. Thanks!

wight-airmash commented:
Haven't looked at the code yet. Just played a couple of times, that's what I noticed:
1. There were no shields spawn in the original BTR (only random inferno https://youtu.be/IP9r-VlwfEM?t=905 and maybe periodic Europe spawn).
2. The players couldn't switch to spectate mode during the active game. Only two ways to finish the game: to win (and get +0 deaths to the stats) or to die (get +1 death).

But I didn't play much of the original, so I could be wrong.

congratulatio commented:
Thanks for the feedback! After confirming in the proxy logs, I've fixed the shield spawn in https://github.com/wight-airmash/ab-server/pull/77/commits/16ef0679d8dc875b4f768b9f24dae9eecffc0983 and removed the periodic inferno in https://github.com/wight-airmash/ab-server/pull/77/commits/aca1d0d13bd1b10bb897cd6ee2d91710ea2e63f1.

congratulatio commented:
I found an excerpt from one of the logs ([original](https://dev.azure.com/airmash/21d867c9-db48-444a-9bd6-015dcc7db5f8/_apis/git/repositories/9f7a1512-eb51-44ec-bfe2-689433160751/items?path=%2Fbtr1%2Fus%2F20190423003258073-518cf5b6d1dadb08-login-919828a265e14ccc81edc22edf520bda.log&versionDescriptor%5BversionOptions%5D=0&versionDescriptor%5BversionType%5D=0&versionDescriptor%5Bversion%5D=room-region&resolveLfs=true&%24format=octetStream&api-version=5.0&download=true), [converted to text](https://github.com/wight-airmash/ab-server/files/4119502/us-20190423003258073-518cf5b6d1dadb08-login-919828a265e14ccc81edc22edf520bda.txt)) showing an in-match spectate request:

```
2019-04-23T00:46:51.619Z <= PING_RESULT (6) {"ping":79,"playerstotal":0,"playersgame":2}
2019-04-23T00:46:52.004Z <= GAME_FIREWALL (33) {"type":1,"status":1,"posX":987.5,"posY":-1967,"radius":15231.76953125,"speed":-70}
2019-04-23T00:46:52.548Z <= PLAYER_UPDATE (12) {"clock":2214292206,"id":41451,"keystate":0,"upgrades":0,"posX":6016.44921875,"posY":502.59765625,"rot":0.393829345703125,"speedX":0.0006103515625,"speedY":0.001220703125}
2019-04-23T00:46:52.622Z => COMMAND (11) {"com":"spectate","data":"-3"}
2019-04-23T00:46:52.623Z <= PLAYER_KILL (17) {"id":41451,"killer":0,"posX":0,"posY":0}
2019-04-23T00:46:52.623Z <= GAME_SPECTATE (31) {"id":41452}
2019-04-23T00:46:52.991Z <= PLAYER_UPDATE (12) {"clock":2214336438,"id":41452,"keystate":0,"upgrades":0,"posX":-1395.408203125,"posY":-4644.880859375,"rot":0,"speedX":0,"speedY":0}
2019-04-23T00:46:53.011Z <= SERVER_CUSTOM (91) {"type":1,"data":"{\"p\":\"[41452]\",\"f\":11,\"b\":1000,\"k\":0,\"t\":13}"}
2019-04-23T00:46:53.011Z <= GAME_FIREWALL (33) {"type":1,"status":0,"posX":987.5,"posY":-1967,"radius":15196.51171875,"speed":-70}
2019-04-23T00:46:53.011Z <= GAME_PLAYERSALIVE (32) {"players":1}
2019-04-23T00:46:53.011Z <= SERVER_MESSAGE (90) {"type":1,"duration":5000,"text":"Game ended!"}
```

Two players in a match with firewall active, player 41451 requests spectate, which is granted and causes 41451 to forfeit the match, with player 41452 declared the winner.

congratulatio commented:
Thanks for the merge. I've added BTR servers in Europe and US (https://github.com/airmash-refugees/airmash-games/commit/863683beb0d6cc489280d8affaf1ee7a162c03c1) and am running one of the FFAs from the same code, seems to be working alright so far. 

The only change since is implementing AFK disconnection, to avoid players passively gaming the points/levels system, I will make a PR for that soon after refactoring.

wight-airmash commented on src/server/modes/base/maintenance/spectating.ts:
```diff
@@ -138,16 +139,29 @@ export default class GameSpectating extends System {
+
+      spectator.spectate.current = playerId;
+      this.subscribeToViewport(playerId, spectatorId);
+
```
I'm not sure about this behavior regarding FFA and CTF. I remember for sure that in CTF after switching to spectate mode, I could stay to observe the area of the map I was on at the moment of switching to spectate mode without auto assignment the player to spectate.

But I would like someone else to confirm it, then I would add a condition only for BTR.

congratulatio replied:
I do not know what the behaviour was of CTF, but I think on FFA it would pick a random player.

I'll analyse this from the proxy logs and let you know what I find.

wight-airmash replied:
Now I think it's better to leave your implementation. This switching logic has several advantages: it reduces the number of objects in the collision system, and it complicates the use of timers in CTF a bit.

Also I'm going to add the feature proposed by xplay: in CTF to switch to the player who currently has taken the flag, i.e. to sort `playerIds` depending on the game state and mode.

I suppose the same adapts for FFA: to switch to the player who was the last killer of the switcher, and to sort other players by the bounty value.

congratulatio replied:
Agreed that feature will be an improvement over the current behaviour. As the implementation of the original game servers will not be copied, I won't continue with the proxy logs analysis.
